### DISCUSSION TITLE
Remove dev requirements for local registry setup

### DISCUSSION THREAD

**Discussion by roee88** [fybrik/fybrik#301] [Developers] [answered]
As part of testing a local kind registry is set up. Currently:
1. Certificates are generated and configured because helm can't push to insecure registries https://github.com/helm/helm/issues/6324. Our current script works in Ubuntu and Mac but not in other linux distros and only when setting GODEBUG=x509ignoreCN=0 environment variable.
2. The developer needs to modify `/etc/hosts` in his own machine

We can keep it as is and just document the requirements better. However, I would like to consider alternatives that would remove the requirements from developers to set `GODEBUG=x509ignoreCN=0` and to modify `/etc/hosts`. Also, removing the burden of maintaining support for different Linux distros in the certificates generation script would be nice (we push away a dev running CentOS for example). See related issue https://github.com/IBM/the-mesh-for-data/issues/287.

**Comment by roee88** [comment]
One alternative is to always use `localhost:5000` as the registry URL just like in the instructions in https://kind.sigs.k8s.io/docs/user/local-registry/ and https://helm.sh/docs/topics/registries/. This removes all of the dev requirements described above.

The main issue is that the `manager` pod will try to pull from `localhost:5000` and fail because in the docker network the registry actually runs in `kind-registry:5000`. We can bypass it by patching the manager with an init container that runs `socat` to redirect localhost:5000 to kind-registry:5000 as part of the integration-tests kustomization layer.

Opinions?

**Reply by froesef** [reply]
If the registry is running with SSL (as it currently is) pushing the charts with helm will not work as helm will always use http instead of https to push to a host that is called `localhost`. (Last time I tried but maybe something got patched). 
I think the other direction (if host is not `localhost` it wants to use https) I think this is why we are in this dilemma. So what could help is running the registry to the outside without SSL and somehow inject a proxy in K8s that introduces SSL might be another solution.

**Reply by roee88** [reply]
This suggestion is to remove using ssl. You always run on localhost without certificates just like in the linked guides. The use of socat is for allowing connecting to a localhost:5000 endpoint from within the manager pod (over http)

**Reply by roee88** [reply]
I'm going to implement it unless there is any objection

**Reply by shlomitk1** [reply]
Possibly related to this issue. Is there a way to test using localhost without having sudo permissions on the machine?

**Reply by roee88** [reply]
Did you need sudo for anything other than modifying /etc/hosts and installing the local registry certificates? For example, does docker require sudo? If so we can check podman.

**Reply by shlomitk1** [reply]
No, just the registry setup.

**Comment by roee88** [comment]
Another alternative is to read the local-registry-hosting configmap (see https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#specification-for-localregistryhosting-v1 and https://kind.sigs.k8s.io/docs/user/local-registry/#create-a-cluster-and-registry) in manager and when pulling from a registry that is listed as the `host`, pull from `hostFromClusterNetwork` instead. 

This won't remove the need for our script to setup certificates (until https://github.com/helm/helm/issues/6324 is fixed) so we will need to find a better way to generate these certificates. For example using https://github.com/FiloSottile/mkcert.

**Comment by roee88** [comment, marked as answer]
Issue opened https://github.com/IBM/the-mesh-for-data/issues/310
